src/main/java/com/reshiftsecurity/plugins/intellij/service/AnalyticsService.java

```diff
@@ -0,0 +1,195 @@
+    private final String PROTOCOL_VERSION = "1";
+    private final String PROTOCOL_VERSION_KEY = "v";
+    private final String HIT_TYPE = "event";
+    private final String HIT_TYPE_KEY = "t";
+    private final String EVENT_ACTION_KEY = "ea";
+    private final String USER_ID_KEY = "cid";
+    private final String MEASUREMENT_ID_KEY = "tid";
+    private final String ACTION_VALUE_KEY = "cm1";
+    private final String EVENT_VALUE_KEY = "ev";
+    private final String ACTION_CATEGORY_KEY = "ec";
+    private final String ACTION_LABEL_KEY = "el";
+    private final String DOC_PATH = "%2Fintellij";
+    private final String DOC_PATH_KEY = "dp";
+    private final String USER_AGENT = "Reshift Intellij IDE Plugin";
+
+    private String userID;
+    private String applicationVersion;
+    private String measurementID;
+    private String sourceKeywords;
+
+    public AnalyticsService() {
+        this.entries = new ArrayList<>();
+        this.applicationVersion = VersionManager.getVersion();
+        this.userID = getUserIdentifier();
+        this.measurementID = "UA-149586212-2";
+        this.sourceKeywords = System.getProperty("os.name") + " " + USER_AGENT + " " + VersionManager.getVersion();
+    }
+
+    public static AnalyticsService getInstance() {
+        return ServiceManager.getService(AnalyticsService.class);
+    }
+
+    public void recordConsentAction(boolean consent) {
+        AnalyticsAction action = consent ? AnalyticsAction.SETTINGS_GATHER_DATA_YES : AnalyticsAction.SETTINGS_GATHER_DATA_NO;
+        this.entries.add(new AnalyticsEntry(action));
+        this.processActions(true);
+    }
+
+    public void recordAction(AnalyticsAction action, String label) {
+        this.entries.add(new AnalyticsEntry(action, null, label));
+        this.processActions();
+    }
+
+    public void recordAction(AnalyticsAction action) {
+        this.entries.add(new AnalyticsEntry(action));
+        this.processActions();
+    }
+
+    public void recordMetric(AnalyticsAction action, Integer value) {
+        this.entries.add(new AnalyticsEntry(action, value));
+        this.processActions();
+    }
+
+    private static String bytesToHex(byte[] hash) {
+        StringBuffer hexString = new StringBuffer();
+        for (int i = 0; i < hash.length; i++) {
+            String hex = Integer.toHexString(0xff & hash[i]);
+            if(hex.length() == 1) hexString.append('0');
+            hexString.append(hex);
+        }
+        return hexString.toString();
+    }
+
+    private String getUserIdentifier() {
+        try {
+            InetAddress ip = InetAddress.getLocalHost();
+            NetworkInterface network = NetworkInterface.getByInetAddress(ip);
+            byte[] mac = network.getHardwareAddress();
+
+            StringBuilder sb = new StringBuilder();
+            for (int i = 0; i < mac.length; i++) {
+                sb.append(String.format("%02X%s", mac[i], (i < mac.length - 1) ? "-" : ""));
```

Comment:
if this is going to be a hash to identify the user, why do you need the '-' in between the hex values?

Reply:
From the docs it seemed this is the format they are expecting: https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters#cid

Reply:
sorry I don't understand how this is the same, the mac address, in this context is formatted as xx-xx-xx-xx-xx-xx, but after this, the sb is hashed by sha-256, and then returned. So in this sense adding "-" doesn't really do anything

The documentation that you outlined says the format of the cid
cid=35009a79-1a05-49d7-b876-2b884d0f825b

but I'm not sure how the return of this method gets put into the cid value

Reply:
oh good catch! lol yep you're right. The hyphens are not needed anymore lol will remove :)

Reply:
try checking if you can fetch a user and their data, I'd explore google bigquery to see if you can get all elements for that user identifier. If I was google, and a client doesn't conform to those standards, I'd silently fail 😱

Reply:
ended up using "uid" which gives more freedom for the format: https://developers.google.com/analytics/devguides/collection/protocol/v1/parameters#uid